File: src/main/kotlin/magicbook/gtlitecore/client/renderer/handler/DimensionDisplayItemRenderer.java
package magicbook.gtlitecore.client.renderer.handler;

import codechicken.lib.model.ModelRegistryHelper;
import codechicken.lib.render.item.IItemRenderer;
import magicbook.gtlitecore.client.model.IWrappedModelGetter;
import magicbook.gtlitecore.common.block.itemblocks.DimensionDisplayItemBlock;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.FontRenderer;
import net.minecraft.client.renderer.RenderItem;
import net.minecraft.client.renderer.block.model.IBakedModel;
import net.minecraft.client.renderer.block.model.ItemCameraTransforms;
import net.minecraft.item.ItemStack;
import net.minecraftforge.common.model.IModelState;
import org.lwjgl.opengl.GL11;

public class DimensionDisplayItemRenderer implements IItemRenderer
{

    protected IBakedModel wrapped;
    private final IModelState state;

    public DimensionDisplayItemRenderer(IModelState state,
                                        IWrappedModelGetter getter)
    {
        this.state = state;
        ModelRegistryHelper.registerPreBakeCallback(modelRegistry -> wrapped = getter.getWrappedModel(modelRegistry));
    }

    @Override
    public void renderItem(ItemStack stack,
                           ItemCameraTransforms.TransformType transformType)
    {
        if (stack.getItem() instanceof DimensionDisplayItemBlock)
        {
            RenderItem renderItem = Minecraft.getMinecraft().getRenderItem();

            String dimension = DimensionDisplayItemBlock.getDimension(stack);
            if (dimension == null || transformType != ItemCameraTransforms.TransformType.GUI)
                return;

            renderItem.renderItemIntoGUI(stack, 0, 0);

            FontRenderer fontRenderer = Minecraft.getMinecraft().fontRenderer;
            float smallTextScale = 3F / 4F;

            GL11.glPushMatrix();
            GL11.glTranslatef(0, 0, 300);
            GL11.glScalef(smallTextScale, smallTextScale, 1.0f);

            long prefix = getPrefix(dimension);
            String tooltipPrefix = prefix != -1 ? "T" + prefix : "ERROR";
            fontRenderer.drawString(tooltipPrefix, 0,
                    (int) (16 / smallTextScale) - fontRenderer.FONT_HEIGHT + 1,
                    0xFFFFFF, true);

            GL11.glPopMatrix();
            GL11.glDisable(GL11.GL_ALPHA_TEST);
        }
    }

    public static long getPrefix(String dimName)
    {
        switch (dimName)
        {
            case "Ow":
                return 1;
            case "Ne":
                return 2;
            case "ED":
                return 3;
        }
        return 0;
    }

    @Override
    public IModelState getTransforms()
    {
        return state;
    }

    @Override
    public boolean isAmbientOcclusion()
    {
        return false;
    }

    @Override
    public boolean isGui3d()
    {
        return false;
    }

}

File: src/main/kotlin/gregtechlite/gtlitecore/client/renderer/handler/DimensionDisplayItemRenderer.kt
package gregtechlite.gtlitecore.client.renderer.handler

import codechicken.lib.model.ModelRegistryHelper
import codechicken.lib.render.CCModelState
import codechicken.lib.render.item.IItemRenderer
import codechicken.lib.util.TransformUtils
import gregtechlite.magicbook.client.Games
import gregtechlite.gtlitecore.common.block.item.DimensionDisplayItemBlock
import net.minecraft.client.renderer.block.model.IBakedModel
import net.minecraft.client.renderer.block.model.ItemCameraTransforms
import net.minecraft.client.renderer.block.model.ModelResourceLocation
import net.minecraft.item.ItemStack
import org.lwjgl.opengl.GL11

class DimensionDisplayItemRenderer(location: ModelResourceLocation) : IItemRenderer
{

    private lateinit var model: IBakedModel

    init
    {
        ModelRegistryHelper.registerPreBakeCallback {
            model = it.getObject(location) ?: throw IllegalArgumentException("Model not found")
        }
    }

    companion object
    {
        fun getPrefix(dimName: String) = when (dimName)
        {
            "Ow" -> 1L
            "Ne" -> 2L
            "ED" -> 3L
            else -> 0L
        }
    }

    override fun renderItem(stack: ItemStack,
                            transformType: ItemCameraTransforms.TransformType?)
    {
        if (stack.item is DimensionDisplayItemBlock)
        {
            val renderItem = Games.renderItem()

            val dimension = DimensionDisplayItemBlock.Companion.getDimension(stack)
            if (dimension == null || transformType != ItemCameraTransforms.TransformType.GUI) return

            renderItem.renderItemIntoGUI(stack, 0, 0)

            val fontRenderer = Games.fontRenderer()
            val smallTextScale = 3f / 4f

            GL11.glPushMatrix()
            GL11.glTranslatef(0f, 0f, 300f)
            GL11.glScalef(smallTextScale, smallTextScale, 1.0f)

            val prefix = getPrefix(dimension)
            val tooltipPrefix = if (prefix != -1L) "T$prefix" else "ERROR"
            fontRenderer.drawString(tooltipPrefix, 0f,
                ((16 / smallTextScale).toInt() - fontRenderer.FONT_HEIGHT + 1).toFloat(),
                0xFFFFFF, true)

            GL11.glPopMatrix()
            GL11.glDisable(GL11.GL_ALPHA_TEST)
        }
    }

    override fun getTransforms(): CCModelState? = TransformUtils.DEFAULT_ITEM

    override fun isAmbientOcclusion() = false

    override fun isGui3d() = false

}